KAFKA-14912:Add a dynamic config for remote index cache size

hudeqi opened the pull request:
As https://issues.apache.org/jira/browse/KAFKA-14912 described, add a dynamic config for remote index cache size.

Commit:
KAFKA-14912:Add a dynamic config for remote index cache size

Commit:
fix conlict

Commit:
fix

hudeqi commented:
Hi, @divijvaidya @satishd If you have time, help to review this pr, thanks.

Commit:
fix conflict

divijvaidya commented on storage/src/main/java/org/apache/kafka/server/log/remote/storage/RemoteLogManagerConfig.java:
```diff
@@ -370,6 +384,10 @@ public long remoteLogIndexFileCacheTotalSizeBytes() {
         return remoteLogIndexFileCacheTotalSizeBytes;
     }
 
+    public long getRemoteLogIndexFileCacheTotalSize() {
```
In Kafka code base we don't use prefix such as "get" for getters. You can rename this to `remoteLogIndexFileCacheTotalSize()`

divijvaidya reviewed (commented):
Continuing discussion on Jira at https://issues.apache.org/jira/browse/KAFKA-14912?focusedCommentId=17756886&page=com.atlassian.jira.plugin.system.issuetabpanels%3Acomment-tabpanel#comment-17756886 

We would prefer to have the cache eviction based on size since based on number of entries is not deterministic.

Commit:
use Caffeine weigher

Commit:
Merge branch 'trunk' of https://github.com/hudeqi/kafka into hdq_remote_index_cache

Commit:
fix unit test

hudeqi commented:
> Continuing discussion on Jira at https://issues.apache.org/jira/browse/KAFKA-14912?focusedCommentId=17756886&page=com.atlassian.jira.plugin.system.issuetabpanels%3Acomment-tabpanel#comment-17756886
> 
> We would prefer to have the cache eviction based on size since based on number of entries is not deterministic.

Hi, I have updated the cache to use weigher to evict based on bytes size. @divijvaidya

divijvaidya commented on storage/src/main/java/org/apache/kafka/storage/internals/log/RemoteIndexCache.java:
```diff
@@ -669,4 +689,10 @@ public static String remoteTransactionIndexFileName(RemoteLogSegmentMetadata rem
     }
 
+    public static int estimatedEntrySize(Entry entry) {
+        return entry.offsetIndex.sizeInBytes() + (int) entry.offsetIndex.length() +
```
why are we double counting offset and time index?

hudeqi replied:
Sorry, I mistook mmap and file as two different things.

divijvaidya commented on storage/src/main/java/org/apache/kafka/storage/internals/log/RemoteIndexCache.java:
```diff
@@ -669,4 +689,10 @@ public static String remoteTransactionIndexFileName(RemoteLogSegmentMetadata rem
     }
 
+    public static int estimatedEntrySize(Entry entry) {
+        return entry.offsetIndex.sizeInBytes() + (int) entry.offsetIndex.length() +
```
"Entry" is immutable so the size won't change over the lifetime of this instance. Can we calculate the size once when we fetch it (in the constructor) and then store it in a member variable? This will prevent calculating sizes every time eviction algorithm runs.

hudeqi replied:
great! updated.

divijvaidya commented on storage/src/main/java/org/apache/kafka/storage/internals/log/RemoteIndexCache.java:
```diff
@@ -437,7 +457,7 @@ public void close() {
                 // Note that internal cache does not require explicit cleaning/closing. We don't want to invalidate or cleanup
                 // the cache as both would lead to triggering of removal listener.
-
+                internalCache.cleanUp();
```
this is not required since GC takes care of clean up. In fact, cleanUp is expensive because it performs maintenance activity such as eviction etc which we don't want to do.

hudeqi replied:
updated

divijvaidya commented on storage/src/main/java/org/apache/kafka/storage/internals/log/RemoteIndexCache.java:
```diff
@@ -125,12 +126,37 @@ public RemoteIndexCache(RemoteStorageManager remoteStorageManager, String logDir
+    public void resizeCacheSize(long remoteLogIndexFileCacheSize) {
+        lock.writeLock().lock();
+        try {
+            Cache<Uuid, Entry> newCache = initEmptyCache(remoteLogIndexFileCacheSize);
+            for (Map.Entry<Uuid, Entry> entry : internalCache.asMap().entrySet()) {
+                newCache.put(entry.getKey(), entry.getValue());
+            }
+            internalCache = newCache;
```
This could get messy as the entries are now present in two caches for some time and old one may try to get rid of them before going out of scope.

I would suggest to keep an expectation that resized cache always starts as empty. We don't need to eagerly populate it since resize is a rare operation.

Please also add a comment here about the reason why we aren't populating the new cache. It will help future readers make sense of the decisions we made in this code change.

hudeqi replied:
updated

divijvaidya reviewed (commented):
Haven't looked at all files yet but leaving some initial review comments

Commit:
fix

Commit:
remove unused import

divijvaidya commented on storage/src/main/java/org/apache/kafka/storage/internals/log/RemoteIndexCache.java:
```diff
@@ -669,4 +705,8 @@ public static String remoteTransactionIndexFileName(RemoteLogSegmentMetadata rem
     }
 
+    public static int estimatedEntrySize(Entry entry) {
+        return entry.offsetIndex.sizeInBytes() + entry.timeIndex.sizeInBytes() + (int) entry.txnIndex.file().length();
```
please use https://docs.oracle.com/javase/8/docs/api/java/nio/file/Files.html#size-java.nio.file.Path- as it is more performant that calculating the length()

divijvaidya replied:
I should have been more clear here. Please keep `entry.offsetIndex.sizeInBytes() + entry.timeIndex.sizeInBytes()` since this is the most efficient way to calculate size since it's in-memory computation. Please use Files.size() instead of `(int) entry.txnIndex.file().length();` since txnIndex does not have a `sizeInBytes` method similar to other two.

showuon replied:
Please change to:
`return offsetIndex.sizeInBytes() + timeIndex.sizeInBytes() + Files.size(txnIndex.file().toPath());`

as @divijvaidya 's suggestion. The reason is when using `sizeInBytes`, we don't need read disk.

divijvaidya commented on storage/src/main/java/org/apache/kafka/storage/internals/log/RemoteIndexCache.java:
```diff
@@ -669,4 +705,8 @@ public static String remoteTransactionIndexFileName(RemoteLogSegmentMetadata rem
         return generateFileNamePrefixForIndex(remoteLogSegmentMetadata) + LogFileUtils.TXN_INDEX_FILE_SUFFIX;
     }
 
+    public static int estimatedEntrySize(Entry entry) {
```
this should return long. Sum of individual ints could be long.

hudeqi replied:
I have updated it, but the embarrassing thing is: even if long is returned here, the weighter implementation in Caffeine cache still needs to convert long to int, because the return type of weigher is int. @divijvaidya

divijvaidya commented on core/src/test/scala/unit/kafka/log/remote/RemoteIndexCacheTest.scala:
```diff
@@ -64,7 +64,7 @@ class RemoteIndexCacheTest {
       time.milliseconds(), brokerId, time.milliseconds(), segmentSize, Collections.singletonMap(0, 0L))
 
-    cache = new RemoteIndexCache(rsm, tpDir.toString)
+    cache = new RemoteIndexCache(Long.MaxValue, rsm, tpDir.toString)
```
Long.MaxValue is a very large size to initialize with. It will cause OOM. Can we have it to a smaller value here which would be sufficient for number of entries we are adding in the tests such as 1 MB?

divijvaidya commented on storage/src/main/java/org/apache/kafka/storage/internals/log/RemoteIndexCache.java:
```diff
@@ -463,6 +489,8 @@ public static class Entry implements AutoCloseable {
 
         private boolean markedForCleanup = false;
 
+        private int entrySize = 0;
```
suggested rename entrySizeBytes

divijvaidya commented on storage/src/main/java/org/apache/kafka/storage/internals/log/RemoteIndexCache.java:
```diff
@@ -395,7 +418,11 @@ private RemoteIndexCache.Entry createCacheEntry(RemoteLogSegmentMetadata remoteL
 
-            return new Entry(offsetIndex, timeIndex, txnIndex);
+            Entry entry = new Entry(offsetIndex, timeIndex, txnIndex);
+            int entrySize = estimatedEntrySize(entry);
+            entry.setEntrySize(entrySize);
```
instead of exposing a public function to set the entry size, perhaps it will be better to keep it private and calculate in constructor. Asking because there is no use case where an entity outside `Entry` will ever possibly want to set entry size based on it's own calculation.

divijvaidya commented on storage/src/main/java/org/apache/kafka/storage/internals/log/RemoteIndexCache.java:
```diff
@@ -110,12 +110,12 @@ public class RemoteIndexCache implements Closeable {
 
     public RemoteIndexCache(RemoteStorageManager remoteStorageManager, String logDir) throws IOException {
-        this(1024, remoteStorageManager, logDir);
+        this(1024L, remoteStorageManager, logDir);
```
please move this into a constant, DEFAULT_REMOTE_INDEX_CACHE_SIZE_BYTES

divijvaidya commented on core/src/main/scala/kafka/server/DynamicBrokerConfig.scala:
```diff
@@ -1125,3 +1128,48 @@ class DynamicProducerStateManagerConfig(val producerStateManagerConfig: Producer
 
 }
+
+class DynamicRemoteLogManagerConfig(server: KafkaBroker) extends BrokerReconfigurable with Logging {
```
Perhaps keep it is `DynamicRemoteLogConfig` so that we can add other remote storage related configs here. This will also align with RemoteLogConfig sub-class in LogConfig.java

divijvaidya commented on core/src/test/scala/unit/kafka/log/remote/RemoteIndexCacheTest.scala:
```diff
@@ -406,9 +417,16 @@ class RemoteIndexCacheTest {
   @Test
   def testReloadCacheAfterClose(): Unit = {
     // close existing cache created in test setup before creating a new one
+    val cacheForEstimate = new RemoteIndexCache(2L, rsm, tpDir.toString)
+    val tpIdForEstimate = new TopicIdPartition(Uuid.randomUuid(), new TopicPartition("foo", 0))
+    val metadataListForEstimate = generateRemoteLogSegmentMetadata(size = 1, tpIdForEstimate)
+    val entryForEstimate = cacheForEstimate.getIndexEntry(metadataListForEstimate.head)
+    val estimateEntryBytesSize = RemoteIndexCache.estimatedEntrySize(entryForEstimate)
+    Utils.closeQuietly(cacheForEstimate, "RemoteIndexCache created for estimating entry size")
```
do we need to do all this? Instead can't we initialize with a big enough size that works for this test case?

hudeqi replied:
Because this test will test the size reserved for only two elements, that is, put three in, it will be verified that since there are only two in the eviction, the cache here cannot be set large enough. @divijvaidya

Commit:
fix

hudeqi commented:
reminding here @divijvaidya

divijvaidya commented on storage/src/main/java/org/apache/kafka/storage/internals/log/RemoteIndexCache.java:
```diff
@@ -77,6 +77,7 @@ public class RemoteIndexCache implements Closeable {
     private static final String TMP_FILE_SUFFIX = ".tmp";
     public static final String REMOTE_LOG_INDEX_CACHE_CLEANER_THREAD = "remote-log-index-cleaner";
     public static final String DIR_NAME = "remote-log-index-cache";
+    public static final long DEFAULT_REMOTE_INDEX_CACHE_SIZE_BYTES = 1024 * 1024L;
```
do we need this at public access?

hudeqi replied:
No, updated.

divijvaidya commented on storage/src/main/java/org/apache/kafka/storage/internals/log/RemoteIndexCache.java:
```diff
@@ -110,12 +111,12 @@ public class RemoteIndexCache implements Closeable {
      * We use {@link Caffeine} cache instead of implementing a thread safe LRU cache on our own.
      */
-    private final Cache<Uuid, Entry> internalCache;
+    private Cache<Uuid, Entry> internalCache;
```
please move this below all the `final` member variables so that all final variables are kept clubbed together. Helps in readibility. Please don't forget to move the java doc along.

showuon commented:
I'll take a look this week. Thanks for the PR.

divijvaidya reviewed (commented):
I am yet to look into the part about why we need `@volatile var remoteLogManagerOpt` and why we need to modify tests but leaving some intermediate comments meanwhile.

showuon commented on storage/src/main/java/org/apache/kafka/storage/internals/log/RemoteIndexCache.java:
```diff
@@ -125,12 +126,39 @@ public RemoteIndexCache(RemoteStorageManager remoteStorageManager, String logDir
      * @param logDir               log directory
      */
-    public RemoteIndexCache(int maxSize, RemoteStorageManager remoteStorageManager, String logDir) throws IOException {
+    public RemoteIndexCache(long maxSize, RemoteStorageManager remoteStorageManager, String logDir) throws IOException {
```
Please update the javadoc for `maxSize`

hudeqi replied:
updated.

showuon commented on core/src/test/scala/unit/kafka/log/remote/RemoteIndexCacheTest.scala:
```diff
@@ -171,21 +178,24 @@ class RemoteIndexCacheTest {
     // Calling getIndex on the same entry should not call rsm#fetchIndex again, but it should retrieve from cache
     cache.getIndexEntry(metadataList.head)
     assertCacheSize(1)
-    verifyFetchIndexInvocation(count = 1)
+    // contains the 1 time for rsm when estimating entry size
+    verifyFetchIndexInvocation(count = 2)
```
The comment in L178 is not correct now.

showuon replied:
Also, I don't understand why we still need to `fetchIndex` again here? If we still invoke fetchIndex under this situation, then the cache doesn't work as expected.

showuon replied:
Ah, sorry, I missed the comment about the estimating entry size.
In this case, you can call:
`clearInvocations(rsm)` after `cacheForEstimate` is closed.

showuon replied:
So that you can remove all the comments about 
`contains the 1 time for rsm when estimating entry size`

hudeqi replied:
updated

showuon replied:
Looks good now. Thanks.

showuon commented on core/src/test/scala/unit/kafka/log/remote/RemoteIndexCacheTest.scala:
```diff
@@ -171,21 +178,24 @@ class RemoteIndexCacheTest {
     assertCacheSize(2)
-    verifyFetchIndexInvocation(count = 2)
+    // contains the 1 time for rsm when estimating entry size
+    verifyFetchIndexInvocation(count = 3)
```
The comment in L184 should be updated.

showuon commented on core/src/test/scala/unit/kafka/log/remote/RemoteIndexCacheTest.scala:
```diff
@@ -159,9 +159,16 @@ class RemoteIndexCacheTest {
+    val metadataListForEstimate = generateRemoteLogSegmentMetadata(size = 1, tpIdForEstimate)
+    val entryForEstimate = cacheForEstimate.getIndexEntry(metadataListForEstimate.head)
+    val estimateEntryBytesSize = entryForEstimate.entrySizeBytes()
+    Utils.closeQuietly(cacheForEstimate, "RemoteIndexCache created for estimating entry size")
```
This estimation should only do once for this test suite. Could we move them into `beforeAll` ?

hudeqi replied:
"@BeforeAll" in scala may be a bit more complicated to implement here. So I extracted this logic into a separate method. @showuon

showuon replied:
Fine.

showuon commented on storage/src/main/java/org/apache/kafka/storage/internals/log/RemoteIndexCache.java:
```diff
@@ -494,6 +518,19 @@ public boolean isMarkedForCleanup() {
+                log.warn("Error occurred when estimating remote index cache entry bytes size, just set 0 firstly.", e);
+                return 0L;
+            }
+        }
```
I think we should add `readLock` here to avoid getting incorrect size.

hudeqi replied:
Great catch!

Commit:
fix

hudeqi commented:
> I am yet to look into the part about why we need `@volatile var remoteLogManagerOpt` and why we need to modify tests but leaving some intermediate comments meanwhile.

Thanks for your review, @divijvaidya . Adding the volatile flag to `remoteLogManagerOpt` is just consistent with the `_logManager` next to it to increase readability. In fact, I think the volatile flags of these two variables can be removed. Please reconfirm this.

Commit:
fix

showuon commented on core/src/test/scala/unit/kafka/log/remote/RemoteIndexCacheTest.scala:
```diff
@@ -594,4 +575,16 @@ class RemoteIndexCacheTest {
+    val entryForEstimate = cacheForEstimate.getIndexEntry(metadataListForEstimate.head)
+    val estimateEntryBytesSize = entryForEstimate.entrySizeBytes()
+    Utils.closeQuietly(cacheForEstimate, "RemoteIndexCache created for estimating entry size")
+    cleanup()
+    setup()
```
These 2 calls are unnecessary. Why can't we just do `clearInvocations` here?

hudeqi replied:
> These 2 calls are unnecessary. Why can't we just do `clearInvocations` here?

Because I have reused `tpDir` here, if not delete this `tpDir` after estimating the entry size (here, for convenience, I directly call `cleanup()` and `setup()` to reset), it will result in After the new cache in @Test is initialized later, the entry corresponding to `tpDir` is directly loaded, causing `assertCacheSize(0)` to fail.
So I can also remove `cleanup()` and `setup()`, and change it to: use a different `tpDir` to execute the estimating logic. @showuon

showuon replied:
I see. Thanks.

kamalcph replied:
My preference is to have a different partition directory to avoid using the `setUp` and `cleanup` methods which are common to all the tests inside the test:

``` scala
private def estimateOneEntryBytesSize(): Long = {
    val tp = new TopicPartition("estimate-entry-bytes-size", 0)
    val tpId = new TopicIdPartition(Uuid.randomUuid(), tp)
    val tpDir = new File(logDir, tpId.toString)
    Files.createDirectory(tpDir.toPath)
    val rsm = mock(classOf[RemoteStorageManager])
    mockRsmFetchIndex(rsm)
    val cache = new RemoteIndexCache(2L, rsm, tpDir.toString)
    val metadataList = generateRemoteLogSegmentMetadata(size = 1, tpId)
    val entry = cache.getIndexEntry(metadataList.head)
    val entrySizeInBytes = entry.entrySizeBytes()
    Utils.closeQuietly(cache, "RemoteIndexCache created for estimating entry size")
    entrySizeInBytes
}

// We can reuse this method in the setup()
private def mockRsmFetchIndex(rsm: RemoteStorageManager): Unit = {
  when(rsm.fetchIndex(any(classOf[RemoteLogSegmentMetadata]), any(classOf[IndexType])))
    .thenAnswer(ans => {
      val metadata = ans.getArgument[RemoteLogSegmentMetadata](0)
      val indexType = ans.getArgument[IndexType](1)
      val offsetIdx = createOffsetIndexForSegmentMetadata(metadata)
      val timeIdx = createTimeIndexForSegmentMetadata(metadata)
      val txnIdx = createTxIndexForSegmentMetadata(metadata)
      maybeAppendIndexEntries(offsetIdx, timeIdx)
      indexType match {
        case IndexType.OFFSET => new FileInputStream(offsetIdx.file)
        case IndexType.TIMESTAMP => new FileInputStream(timeIdx.file)
        case IndexType.TRANSACTION => new FileInputStream(txnIdx.file)
        case IndexType.LEADER_EPOCH => // leader-epoch-cache is not accessed.
        case IndexType.PRODUCER_SNAPSHOT => // producer-snapshot is not accessed.
      }
    })
}
```

showuon reviewed (commented):
Thanks for the update. Left a comment. Also, we need to add a test to verify after resizeCacheSize, the cache content will be cleared.

divijvaidya commented on core/src/main/scala/kafka/server/DynamicBrokerConfig.scala:
```diff
@@ -1125,3 +1128,48 @@ class DynamicProducerStateManagerConfig(val producerStateManagerConfig: Producer
+    if (oldValue != newValue) {
+      val remoteLogManager = server.remoteLogManager
+      if (remoteLogManager.nonEmpty) {
+        remoteLogManager.get.resizeCacheSize(newValue)
```
perhaps add a info log here that mentions the old and resized value

divijvaidya commented on core/src/test/scala/unit/kafka/server/DynamicBrokerConfigTest.scala:
```diff
@@ -787,6 +788,31 @@ class DynamicBrokerConfigTest {
   }
 
+  @Test
+  def testUpdateDynamicRemoteLogManagerConfig(): Unit = {
```
Could we also write a test that asserts that the index files on disk (at $logdir/remote-log-index-cache) belonging to older cache gets deleted after resize if done? Otherwise, we may end up with a file leak after resize.

You can create this test at RemoteIndexCacheTest.

hudeqi replied:
There was indeed files leak and I fixed it.

divijvaidya reviewed (commented):
Since we have implemented this config correctly, please make this configuration externally visible at https://github.com/apache/kafka/blob/17862ffaf2ec4efe7e3eeebf70ab9f1e1e1c70c7/storage/src/main/java/org/apache/kafka/server/log/remote/storage/RemoteLogManagerConfig.java#L200 (i.e. change `defineInternal` to `define`)

Commit:
fix

hudeqi commented:
Like the unit test you all mentioned to verify that the cache contents are cleared after resizing, I have fixed and updated it, thanks! @showuon @divijvaidya

ben-manes commented on storage/src/main/java/org/apache/kafka/storage/internals/log/RemoteIndexCache.java:
```diff
@@ -110,27 +113,54 @@ public class RemoteIndexCache implements Closeable {
+
+    public void resizeCacheSize(long remoteLogIndexFileCacheSize) {
+        lock.writeLock().lock();
+        try {
+            // When resizing the cache, we always start with an empty cache. There are two main reasons:
+            // 1. Resizing the cache is not a high-frequency operation, and there is no need to fill the data in the old
+            // cache to the new cache in time when resizing inside.
+            // 2. Since the eviction of the caffeine cache is cleared asynchronously, it is possible that after the entry
+            // in the old cache is filled in the new cache, the old cache will clear the entry, and the data in the two caches
+            // will be inconsistent.
+            internalCache.invalidateAll();
+            log.info("Invalidated all entries in the cache and triggered the cleaning of all index files in the cache dir.");
+            internalCache = initEmptyCache(remoteLogIndexFileCacheSize);
```
fyi, you could [dynamically resize](https://www.javadoc.io/doc/com.github.ben-manes.caffeine/caffeine/latest/com.github.benmanes.caffeine/com/github/benmanes/caffeine/cache/Policy.Eviction.html#setMaximum(@org.checkerframework.checker.index.qual.NonNegativelong)) the cache instance, which will perform the eviction on the calling thread if it exceeds the new maximum.
```java
cache.policy().eviction().orElseThrow().setMaximum(remoteLogIndexFileCacheSize);
```
The [Policy](https://www.javadoc.io/doc/com.github.ben-manes.caffeine/caffeine/latest/com.github.benmanes.caffeine/com/github/benmanes/caffeine/cache/Policy.html) api includes many helpful but ad hoc methods to solve problems, while also letting us try to keep the top-level apis streamlined for the common usages.

showuon replied:
Nice suggestion @ben-manes ! 
@hudeqi , since time is approaching the v3.6.0, I think we could have a follow-up PR to address this comment. WDYT?

hudeqi replied:
I'm working on debugging and adding unit tests to this suggestion. Then according to your suggestion, I separate the commit of this part of the change and put it in another PR? @showuon

showuon replied:
Ah, OK, if you have time, please add the change into this PR. Thanks.

showuon replied:
@hudeqi , sorry, I don't want to add pressure to you, but we're planning to have RC for v3.6.0 release this Wed.. At this point of time, I think we'd better to get the current PR merged, then having a follow-up PR for the cache resize improvement. So that we can have at least a configurable index cache in v3.6.0, even if we can't have the improvement in v3.6.0 in time. 

That said, please keep the current PR as is, and see if @divijvaidya has any other comments. In the mean time, you can create a separate PR for the improvement. Thanks for your effort!

hudeqi replied:
> @hudeqi , sorry, I don't want to add pressure to you, but we're planning to have RC for v3.6.0 release this Wed.. At this point of time, I think we'd better to get the current PR merged, then having a follow-up PR for the cache resize improvement. So that we can have at least a configurable index cache in v3.6.0, even if we can't have the improvement in v3.6.0 in time.
> 
> That said, please keep the current PR as is, and see if @divijvaidya has any other comments. In the mean time, you can create a separate PR for the improvement. Thanks for your effort!

Okay, after this PR is merged, I will put this improved commit into the next PR. grateful! @showuon

Commit:
fix ut

showuon reviewed (approved):
LGTM! Thanks for the patch!

divijvaidya reviewed (approved):
Thank you for you patience with this one @hudeqi. Looks good to be merged. Let's pick up @ben-manes suggestion in a follow up PR. Thank you chiming in the review @ben-manes, you are awesome!

divijvaidya commented:
@hudeqi we have a few TS related tests failing such as https://ci-builds.apache.org/job/Kafka/job/kafka-pr/job/PR-14243/14/testReport/junit/org.apache.kafka.tiered.storage.integration/EnableRemoteLogOnTopicTest/Build___JDK_17_and_Scala_2_13___executeTieredStorageTest_String__quorum_kraft/ 

Can you please check on them?

hudeqi commented:
> @hudeqi we have a few TS related tests failing such as https://ci-builds.apache.org/job/Kafka/job/kafka-pr/job/PR-14243/14/testReport/junit/org.apache.kafka.tiered.storage.integration/EnableRemoteLogOnTopicTest/Build___JDK_17_and_Scala_2_13___executeTieredStorageTest_String__quorum_kraft/
> 
> Can you please check on them?

The strange thing is that I can pass this test locally. @divijvaidya

Commit:
Merge branch 'trunk' of https://github.com/hudeqi/kafka into hdq_remote_index_cache

hudeqi commented:
> > @hudeqi we have a few TS related tests failing such as https://ci-builds.apache.org/job/Kafka/job/kafka-pr/job/PR-14243/14/testReport/junit/org.apache.kafka.tiered.storage.integration/EnableRemoteLogOnTopicTest/Build___JDK_17_and_Scala_2_13___executeTieredStorageTest_String__quorum_kraft/
> > Can you please check on them?
> 
> The strange thing is that I can pass this test locally. @divijvaidya

I'll try merging the latest trunk branch firstly.

satishd commented:
Thanks @hudeqi for the PR. I will start reviewing it in the next couple of days.

satishd reviewed (commented):
Thanks @hudeqi for the PR. I will start reviewing it in the next couple of days.

kamalcph commented:
@hudeqi 

Can you rebase your branch against trunk instead of merge? With rebase, we can review the changes easily.

hudeqi commented:
> @hudeqi
> 
> Can you rebase your branch against trunk instead of merge? With rebase, we can review the changes easily.

Do you mean to merge these temporary commits into one commit? @kamalcph

Commit:
fix conflict

kamalcph commented:
> > @hudeqi
> > Can you rebase your branch against trunk instead of merge? With rebase, we can review the changes easily.
> 
> Do you mean to merge these temporary commits into one commit? @kamalcph

No, not squashing the commits. If you rebase your branch with trunk, then all your commits will be on the top of the tree which will be easy for the reviewers to review the change-set and follow the comments posted on this PR.

kamalcph commented on core/src/main/scala/kafka/server/BrokerServer.scala:
```diff
@@ -700,4 +700,5 @@ class BrokerServer(
 
   override def boundPort(listenerName: ListenerName): Int = socketServer.boundPort(listenerName)
 
+  override def remoteLogManager: Option[RemoteLogManager] = remoteLogManagerOpt
```
Why do we need this method?

trait `KafkaBroker` already have [remoteLogManagerOpt](https://github.com/apache/kafka/blob/trunk/core/src/main/scala/kafka/server/KafkaBroker.scala#L84) method.

hudeqi replied:
It seems that there was no `remoteLogManagerOpt` when I started this PR, but you added it recently. I will reuse `remoteLogManagerOpt`, thanks!

kamalcph commented on core/src/main/scala/kafka/server/DynamicBrokerConfig.scala:
```diff
@@ -1129,3 +1132,50 @@ class DynamicProducerStateManagerConfig(val producerStateManagerConfig: Producer
+}
+
+object DynamicRemoteLogConfig {
+  val DynamicRemoteLogConfigs = Set(
```
Can we follow uniform naming strategy for readability?

Rename `DynamicRemoteLogConfigs` -> `ReconfigurableConfigs`

hudeqi replied:
updated.

kamalcph commented on core/src/main/scala/kafka/server/DynamicBrokerConfig.scala:
```diff
@@ -89,7 +90,8 @@ object DynamicBrokerConfig {
     SocketServer.ReconfigurableConfigs ++
-    ProducerStateManagerConfig.RECONFIGURABLE_CONFIGS.asScala
+    ProducerStateManagerConfig.RECONFIGURABLE_CONFIGS.asScala ++
+    DynamicRemoteLogConfigs
```
Can we remove the static import and make the usage uniform for readability?

```
DynamicRemoteLogConfig.ReconfigurableConfigs
```

hudeqi replied:
updated.

kamalcph commented on core/src/main/scala/kafka/server/KafkaBroker.scala:
```diff
@@ -95,6 +95,7 @@ trait KafkaBroker extends Logging {
   def brokerTopicStats: BrokerTopicStats
   def credentialProvider: CredentialProvider
   def clientToControllerChannelManager: NodeToControllerChannelManager
+  def remoteLogManager: Option[RemoteLogManager]
```
this is not required, please reuse `remoteLogManagerOpt` method.

hudeqi replied:
updated.

kamalcph commented on core/src/main/scala/kafka/server/KafkaServer.scala:
```diff
@@ -129,7 +129,7 @@ class KafkaServer(
   var logDirFailureChannel: LogDirFailureChannel = _
   @volatile private var _logManager: LogManager = _
-  var remoteLogManagerOpt: Option[RemoteLogManager] = None
+  @volatile var remoteLogManagerOpt: Option[RemoteLogManager] = None
```
Why do we have to mark this as `volatile`?

hudeqi replied:
updated.

kamalcph commented on storage/src/main/java/org/apache/kafka/server/log/remote/storage/RemoteLogManagerConfig.java:
```diff
@@ -197,7 +197,7 @@ public final class RemoteLogManagerConfig {
-                  .defineInternal(REMOTE_LOG_INDEX_FILE_CACHE_TOTAL_SIZE_BYTES_PROP,
+                  .define(REMOTE_LOG_INDEX_FILE_CACHE_TOTAL_SIZE_BYTES_PROP,
                                   LONG,
                                   DEFAULT_REMOTE_LOG_INDEX_FILE_CACHE_TOTAL_SIZE_BYTES,
```
`DEFAULT_REMOTE_LOG_INDEX_FILE_CACHE_TOTAL_SIZE_BYTES` is still referring to 1 GB. 

What is the difference between `DEFAULT_REMOTE_INDEX_CACHE_SIZE_BYTES` and `DEFAULT_REMOTE_LOG_INDEX_FILE_CACHE_TOTAL_SIZE_BYTES` configs?

hudeqi replied:
`DEFAULT_REMOTE_INDEX_CACHE_SIZE_BYTES ` is used in unit test. `DEFAULT_REMOTE_LOG_INDEX_FILE_CACHE_TOTAL_SIZE_BYTES ` is a default value for using actually.

kamalcph commented on storage/src/main/java/org/apache/kafka/storage/internals/log/RemoteIndexCache.java:
```diff
@@ -112,27 +115,54 @@ public class RemoteIndexCache implements Closeable {
-    private final ShutdownableThread cleanerThread;
+    private Cache<Uuid, Entry> internalCache;
 
     public RemoteIndexCache(RemoteStorageManager remoteStorageManager, String logDir) throws IOException {
```
Can we remove this constructor? (it's used only by the test)

hudeqi replied:
I have added "Visible for testing", and it is only used in test.

kamalcph commented on core/src/test/scala/unit/kafka/log/remote/RemoteIndexCacheTest.scala:
```diff
@@ -230,7 +231,7 @@ class RemoteIndexCacheTest {
     Utils.closeQuietly(cache, "RemoteIndexCache created for unit test")
 
-    cache = new RemoteIndexCache(2, rsm, tpDir.toString)
+    cache = new RemoteIndexCache(2L, rsm, tpDir.toString)
```
Why this is `2L` but not `2 * estimateEntryBytesSize`?

hudeqi replied:
good catch, thanks.

kamalcph commented on core/src/test/scala/unit/kafka/log/remote/RemoteIndexCacheTest.scala:
```diff
@@ -524,6 +526,50 @@ class RemoteIndexCacheTest {
+    }
+
+    Utils.closeQuietly(cache, "RemoteIndexCache created for unit test")
+    cache = new RemoteIndexCache(rsm, tpDir.toString)
```
Why do we reinitialise the cache?

hudeqi replied:
updated.

kamalcph reviewed (commented):
Overall LGTM, left minor comments to address. Thanks, @hudeqi for working on this!

kamalcph commented on core/src/test/scala/unit/kafka/log/remote/RemoteIndexCacheTest.scala:
```diff
@@ -598,4 +644,16 @@ class RemoteIndexCacheTest {
     }
   }
+
+  private def getEstimateEntryBytesSize(): Long = {
```
Reposting it since the other comment was marked as resolved:

My preference is to have a different partition directory to avoid using the setUp and cleanup methods which are common to all the tests inside the test:

```java
private def estimateOneEntryBytesSize(): Long = {
    val tp = new TopicPartition("estimate-entry-bytes-size", 0)
    val tpId = new TopicIdPartition(Uuid.randomUuid(), tp)
    val tpDir = new File(logDir, tpId.toString)
    Files.createDirectory(tpDir.toPath)
    val rsm = mock(classOf[RemoteStorageManager])
    mockRsmFetchIndex(rsm)
    val cache = new RemoteIndexCache(2L, rsm, tpDir.toString)
    val metadataList = generateRemoteLogSegmentMetadata(size = 1, tpId)
    val entry = cache.getIndexEntry(metadataList.head)
    val entrySizeInBytes = entry.entrySizeBytes()
    Utils.closeQuietly(cache, "RemoteIndexCache created for estimating entry size")
    entrySizeInBytes
}

// We can reuse this method in the setup()
private def mockRsmFetchIndex(rsm: RemoteStorageManager): Unit = {
  when(rsm.fetchIndex(any(classOf[RemoteLogSegmentMetadata]), any(classOf[IndexType])))
    .thenAnswer(ans => {
      val metadata = ans.getArgument[RemoteLogSegmentMetadata](0)
      val indexType = ans.getArgument[IndexType](1)
      val offsetIdx = createOffsetIndexForSegmentMetadata(metadata)
      val timeIdx = createTimeIndexForSegmentMetadata(metadata)
      val txnIdx = createTxIndexForSegmentMetadata(metadata)
      maybeAppendIndexEntries(offsetIdx, timeIdx)
      indexType match {
        case IndexType.OFFSET => new FileInputStream(offsetIdx.file)
        case IndexType.TIMESTAMP => new FileInputStream(timeIdx.file)
        case IndexType.TRANSACTION => new FileInputStream(txnIdx.file)
        case IndexType.LEADER_EPOCH => // leader-epoch-cache is not accessed.
        case IndexType.PRODUCER_SNAPSHOT => // producer-snapshot is not accessed.
      }
    })
}
```

hudeqi replied:
updated. thanks!

Commit:
minor fix

hudeqi commented:
This PR was closed by me deleting the branch by mistake. It seems that I cannot reopen this PR. :)

hudeqi commented:
test reopen.

hudeqi commented:
Sorry, all. I've reopened a PR [here](https://github.com/apache/kafka/pull/14381)

hudeqi commented:
1
